tools/autograd/templates/variable_factories.h

```diff
@@ -104,7 +104,7 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
       // `N` is the number of the elements in the tensor).
       at::Tensor tensor = ([&]() {
         at::AutoNonVariableTypeMode non_var_type_mode(true);
-        return at::empty(sizes_, at::TensorOptions(options).device(at::kCPU).is_variable(false));
+        return at::empty(sizes_, at::TensorOptions(options).device(at::kCPU));
```

Comment:
Should these be guarded with `AutoNonVariableTypeMode` ?
Is that the right guard to use to perform non-autograd stuff in cpp? Or `NoGradGuard` is the right thing to use?

Reply:
The situation with `AutoNonVariableTypeMode` and `NoGradGuard` is very delicate right now. In the terminal state, these two guards would not be distinguishable. However, because tracing, profiling, and version counter logic lives inside VariableType implementations, these guards have different functions: NoGradGuard has the same function as `no_grad`, but we will still execute VariableType implementations; `AutoNonVariableTypeMode` will bypass VariableType entirely.

Most of the insertions of AutoNonVariableTypeMode in this patch are to solve problems with tracing, so `NoGradGuard` wouldn't actually solve our problem (because you are still going to hit the tracing code.) The most common situation looks something like this:

```
Tensor t = at::empty(...);
// do some raw data access on t, e.g., set storage or mutate data pointer
Variable v = make_variable(t);
```

Without the AutoNonVariableTypeMode guard this will trace incorrectly because `at::empty` now generates an `at::empty` in the trace, but then we fail to trace the raw data access. Previously it would have worked as `at::` would have returned a non-Variable which doesn't trigger tracing.

Reply:
From a new user point of view that write brand new cpp code.
- The `NoGradGuard` should be used just like `no_grad()` in python.
- `AutoNonVariableTypeMode` is an internal thing that should NOT be used. Do like python: use requires_grad(false) + detach() to get what you want.

For existing code: people should not be doing factory functions so this is not BC.

Reply:
Yes, that sounds accurate to me

Reply:
I'm going to beef up the documentation on this guard